client/beefy/src/worker.rs

```diff
@@ -148,22 +162,33 @@ where
 	C: Client<B, BE>,
 	C::Api: BeefyApi<B>,
 {
-	/// Return `true`, if we should vote on block `number`
-	fn should_vote_on(&self, number: NumberFor<B>) -> bool {
-		let best_beefy_block = if let Some(block) = self.best_beefy_block {
-			block
+	/// Return `Some(number)` if we should be voting on block `number` now,
+	/// return `None` if there is no block we should vote on now.
+	fn current_vote_target(&self) -> Option<NumberFor<B>> {
+		let rounds = if let Some(r) = &self.rounds {
+			r
 		} else {
-			debug!(target: "beefy", "🥩 Missing best BEEFY block - won't vote for: {:?}", number);
-			return false
+			error!(target: "beefy", "🥩 No voting round started");
+			return None
 		};
 
-		let target = vote_target(self.best_grandpa_block, best_beefy_block, self.min_block_delta);
-
-		trace!(target: "beefy", "🥩 should_vote_on: #{:?}, next_block_to_vote_on: #{:?}", number, target);
-
+		let best_finalized = *self.best_grandpa_block_header.number();
+		let target = vote_target(
+			best_finalized,
+			self.best_beefy_block.unwrap_or(0u32.into()),
+			*rounds.session_start(),
+			self.min_block_delta,
+		);
 		metric_set!(self, beefy_should_vote_on, target);
 
-		number == target
+		trace!(target: "beefy", "🥩 best finalized: #{:?}, next_block_to_vote_on: #{:?}", best_finalized, target);
+
+		// Don't vote for targets until they've been finalized.
+		if target > best_finalized {
```

Comment:
I don't think this should be necessary, we should rather make sure that `vote_target` returns values between `best_finalized` and `best_beefy` block. We might want it to return `None` in case `session_start` is taken and it's not finalized yet or if the target is already beefy-finalized (i.e. we are up-to-date).

Reply:
Actually because of `min_block_delta` param, voting targets can be blocks not yet finalized. In that case, this check keeps us patient until target is actually finalized by GRANDPA.

Reply:
I see. I think it's worth documenting in the function then. Alternatively I'd suggest to return an `Option` from `vote_target` in case there isn't currently one and encapsulate the entire block-selection logic there.